cpp/src/arrow/compute/exec/window_functions/window_rank.cc

```diff
@@ -0,0 +1,456 @@
+    }
+  }
+}
+
+void WindowRank_Framed_Ref::Eval(RankType rank_type,
+                                 const BitVectorNavigator& rank_key_tie_begins,
+                                 const int64_t* order_by_rank_key,
+                                 const WindowFrames& frames, int64_t* results) {
+  int64_t num_rows = rank_key_tie_begins.bit_count();
+
+  std::vector<int64_t> global_ranks_order_by_rank_key(num_rows);
+  WindowRank_Global_Ref::Eval(rank_type, rank_key_tie_begins,
+                              global_ranks_order_by_rank_key.data());
+
+  std::vector<int64_t> global_ranks(num_rows);
+  if (!order_by_rank_key) {
+    for (int64_t i = 0; i < num_rows; ++i) {
+      global_ranks[i] = global_ranks_order_by_rank_key[i];
+    }
+  } else {
+    for (int64_t i = 0; i < num_rows; ++i) {
+      global_ranks[order_by_rank_key[i]] = global_ranks_order_by_rank_key[i];
+    }
+  }
+
+  for (int64_t frame_index = 0; frame_index < frames.num_frames; ++frame_index) {
+    int64_t current_row_index = frames.first_row_index + frame_index;
+
+    // Compute list of global ranks for all rows within the frame.
+    //
+    // Make sure to include the current row in the frame, even if it lies
+    // outside of the ranges defining its.
+    //
+    std::vector<int64_t> global_ranks_within_frame;
+    bool current_row_included = false;
+    for (int64_t range_index = 0; range_index < frames.num_ranges_per_frame;
+         ++range_index) {
+      int64_t begin = frames.begins[range_index][frame_index];
+      int64_t end = frames.ends[range_index][frame_index];
+      if (!current_row_included && current_row_index < begin) {
+        global_ranks_within_frame.push_back(global_ranks[current_row_index]);
+        current_row_included = true;
+      }
+      for (int64_t row_index = begin; row_index < end; ++row_index) {
+        if (row_index == current_row_index) {
+          current_row_included = true;
+        }
+        global_ranks_within_frame.push_back(global_ranks[row_index]);
+      }
+    }
+    if (!current_row_included) {
+      global_ranks_within_frame.push_back(global_ranks[current_row_index]);
+      current_row_included = true;
+    }
+
+    int64_t rank = 0;
+    for (int64_t frame_row_index = 0;
+         frame_row_index < static_cast<int64_t>(global_ranks_within_frame.size());
+         ++frame_row_index) {
+      switch (rank_type) {
+        case RankType::ROW_NUMBER:
+          // Count the number of rows in the frame with lower global rank.
+          //
+          if (global_ranks_within_frame[frame_row_index] <
+              global_ranks[current_row_index]) {
+            ++rank;
+          }
+          break;
+        case RankType::RANK_TIES_LOW:
+          // Count the number of rows in the frame with lower global rank.
+          //
+          if (global_ranks_within_frame[frame_row_index] <
+              global_ranks[current_row_index]) {
+            ++rank;
+          }
+          break;
```

Comment:
The bodies of these cases are identical. I think that may be correct, but wouldn't it be better to be explicit about that by collapsing them?

```suggestion
        case RankType::ROW_NUMBER:
        case RankType::RANK_TIES_LOW:
          // Count the number of rows in the frame with lower global rank.
          //
          if (global_ranks_within_frame[frame_row_index] <
              global_ranks[current_row_index]) {
            ++rank;
          }
          break;
```